codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/MainSourceGenerator.scala

```diff
@@ -0,0 +1,248 @@
+      .sortBy(_.fqn.name)
+      .collect {
+        case entity: ModelBuilder.EventSourcedEntity => s"new ${entity.fqn.name}(_)"
+        case entity: ModelBuilder.ValueEntity        => s"new ${entity.fqn.name}(_)"
+        case entity: ModelBuilder.ReplicatedEntity   => s"new ${entity.fqn.name}(_)"
+      }
+
+    val serviceRegistrationParameters = services.values.toList
+      .sortBy(_.fqn.name)
+      .collect {
+        case service: ModelBuilder.ActionService => s"new ${service.className}(_)"
+        case view: ModelBuilder.ViewService      => s"new ${view.className}(_)"
+      }
+
+    val registrationParameters = entityRegistrationParameters ::: serviceRegistrationParameters
+    s"""|package $mainClassPackageName
+        |
+        |import com.akkaserverless.scalasdk.AkkaServerless
+        |import org.slf4j.LoggerFactory
+        |${componentImports}
+        |
+        |object $mainClassName {
+        |
+        |  private val log = LoggerFactory.getLogger("$mainClassPackageName.$mainClassName")
+        |
+        |  def createAkkaServerless(): AkkaServerless = {
+        |    // The AkkaServerlessFactory automatically registers any generated Actions, Views or Entities,
+        |    // and is kept up-to-date with any changes in your protobuf definitions.
+        |    // If you prefer, you may remove this and manually register these components in a
+        |    // `AkkaServerless()` instance.
+        |    AkkaServerlessFactory.withComponents(
+        |      ${registrationParameters.mkString(",\n      ")})
+        |  }
+        |
+        |  def main(args: Array[String]): Unit = {
+        |    log.info("starting the Akka Serverless service")
+        |    createAkkaServerless().start()
+        |  }
+        |}
+        |""".stripMargin
+
+  }
+
+  private[codegen] def akkaServerlessFactorySource(mainClassPackageName: String, model: ModelBuilder.Model): String = {
+    val registrations = model.services.values
+      .flatMap {
+        case service: ModelBuilder.EntityService =>
+          model.entities.get(service.componentFullName).toSeq.map {
+            case entity: ModelBuilder.EventSourcedEntity =>
+              s".register(${entity.fqn.name}Provider(create${entity.fqn.name}))"
+            case entity: ModelBuilder.ValueEntity =>
+              s".register(${entity.fqn.name}Provider(create${entity.fqn.name}))"
+            case entity: ModelBuilder.ReplicatedEntity =>
+              s".register(${entity.fqn.name}Provider(create${entity.fqn.name}))"
+          }
+
+        case service: ModelBuilder.ViewService =>
+          List(s".register(${service.providerName}(create${service.className}))")
+
+        case service: ModelBuilder.ActionService =>
+          List(s".register(${service.providerName}(create${service.className}))")
+
+      }
+      .toList
+      .sorted
+
+    val entityImports = model.entities.values.flatMap { ety =>
+      if (ety.fqn.parent.javaPackage != mainClassPackageName) {
```

Comment:
this logic is no longer needed here when we use `generateImports` below, I think

Reply:
it wasn't filtering those, only in `typeName`, but I have implemented that in 94cccf7a5e725516a27f4eb7e150e6949bb8a9f9

Reply:
ah, nice :+1: